cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/RemoteTaskExecutor.java

```diff
@@ -0,0 +1,82 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app;
+
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.api.retry.RetryableException;
+import io.cdap.cdap.api.service.worker.RunnableTaskRequest;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.http.DefaultHttpRequestConfig;
+import io.cdap.cdap.common.internal.remote.RemoteClient;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.common.http.HttpMethod;
+import io.cdap.common.http.HttpRequest;
+import io.cdap.common.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Helper class for executing a {@link RunnableTaskRequest} on a remote worker.
+ */
+public class RemoteTaskExecutor {
+
+  private static final Gson GSON = new Gson();
+  private static final String TASK_WORKER_URL = "/worker/run";
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteTaskExecutor.class);
+  private final RemoteClient remoteClient;
+  private final RetryStrategy retryStrategy;
+
+  @Inject
+  RemoteTaskExecutor(CConfiguration cConf, DiscoveryServiceClient discoveryServiceClient) {
+    this.remoteClient = new RemoteClient(discoveryServiceClient, Constants.Service.TASK_WORKER,
+                                         new DefaultHttpRequestConfig(false),
+                                         Constants.Gateway.INTERNAL_API_VERSION_3);
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, Constants.Service.TASK_WORKER + ".");
+  }
+
+  /**
+   * Sends the {@link RunnableTaskRequest} to a remote worker and returns the result.
+   * Retries sending the request if the workers are busy.
+   *
+   * @param runnableTaskRequest
+   * @return
+   * @throws IOException
+   */
+  public byte[] runTask(RunnableTaskRequest runnableTaskRequest) throws IOException {
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.POST, TASK_WORKER_URL);
+    HttpRequest httpRequest = requestBuilder.withBody(GSON.toJson(runnableTaskRequest)).build();
+    return Retries.callWithRetries(() -> {
+      HttpResponse httpResponse = remoteClient.execute(httpRequest);
+      if (httpResponse.getResponseCode() == HttpResponseStatus.TOO_MANY_REQUESTS.code()) {
+        throw new RetryableException(
+          String.format("Received response code %s for %s", httpResponse.getResponseCode(), runnableTaskRequest));
+      } else if (httpResponse.getResponseCode() != 200) {
+        LOG.debug("Received response code {} for {}", httpResponse.getResponseCode(), runnableTaskRequest);
```

Comment:
So even in case of error, the error body will be returned without signaling it is an error? So the caller would try to use the result as if the call was successful?

Reply:
Good point. Constructing an exception here if the response is not 200